peps/pep-0748.rst

```diff
@@ -0,0 +1,923 @@
+providing a pointer to an array of these ``ALG_ID`` constants. This means that any
+suitable API must allow the Python code to determine which ``ALG_ID`` constants must
+be provided.
+
+Network Security Services (NSS)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+NSS is Mozilla's crypto and TLS library. It's used in Firefox, Thunderbird, and
+as an alternative to OpenSSL in multiple libraries, e.g. curl.
+
+By default, NSS comes with secure configuration of allowed ciphers. On some
+platforms such as Fedora, the list of enabled ciphers is globally configured in
+a system policy. Generally, applications should not modify cipher suites unless
+they have specific reasons to do so.
+
+NSS has both process global and per-connection settings for cipher suites. It
+does not have a concept of :class:`~ssl.SSLContext` like OpenSSL. A
+:class:`~ssl.SSLContext`-like behavior can be easily emulated. Specifically,
+ciphers can be enabled or disabled globally with
+``SSL_CipherPrefSetDefault(PRInt32 cipher, PRBool enabled)``, and
+``SSL_CipherPrefSet(PRFileDesc *fd, PRInt32 cipher, PRBool enabled)`` for a
+connection. The cipher ``PRInt32`` number is a signed 32-bit integer that
+directly corresponds to an registered IANA id, e.g. ``0x1301`` is
+``TLS_AES_128_GCM_SHA256``. Contrary to OpenSSL, the preference order of ciphers
+is fixed and cannot be modified at runtime.
+
+Like Network Framework, NSS has no API for aggregated entries. Some consumers of
+NSS have implemented custom mappings from OpenSSL cipher names and rules to NSS
+ciphers, e.g. ``mod_nss``.
+
+Proposed Interface
+^^^^^^^^^^^^^^^^^^
+
+The proposed interface for the new module is influenced by the combined set of
+limitations of the above implementations. Specifically, as every implementation
+except OpenSSL requires that each individual cipher be provided, there is no
+option but to provide that lowest-common denominator approach.
+
+The simplest approach is to provide an enumerated type that includes a large
+subset of the cipher suites defined for TLS. The values of the enum members will
+be their two-octet cipher identifier as used in the TLS handshake, stored as a
+16 bit integer. The names of the enum members will be their IANA-registered
+cipher suite names.
+
+As of now, the `IANA cipher suite registry
+<https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4>`_
+contains over 320 cipher suites. A large portion of the cipher suites are
+irrelevant for TLS connections to network services. Other suites specify
+deprecated and insecure algorithms that are no longer provided by recent
+versions of implementations. The enum contains the five fixed cipher suites
+defined for TLS v1.3, and for TLS v1.2, it only contains the cipher suites that
+correspond to the TLS v1.3 cipher suites, with ECDHE key exchange (for perfect
+forward secrecy) and ECDSA or RSA signatures, which are an additional ten cipher
+suites.
```

Comment:
Likely worth explicitly stating here that libraries and applications that need interoperability with legacy ciphersuites will need to continue to use the existing OpenSSL based TLS APIs.

Reply:
Or add a sentence referring to the acceptance of raw integers covered in more detail later